File: src/utils/validation.ts
import { PanchangError } from '../types/errors';
import type { GeoLocation } from '../types/location';
import { TOTAL_NAKSHATRAS, TOTAL_TITHIS } from './constants';

export function validateLocation(location: GeoLocation): void {
  if (typeof location.latitude !== 'number' || location.latitude < -90 || location.latitude > 90) {
    throw new PanchangError(
      `Latitude must be a number between -90 and 90, got ${location.latitude}`,
      'INVALID_LATITUDE'
    );
  }
  if (typeof location.longitude !== 'number' || location.longitude < -180 || location.longitude > 180) {
    throw new PanchangError(
      `Longitude must be a number between -180 and 180, got ${location.longitude}`,
      'INVALID_LONGITUDE'
    );
  }
  if (location.elevation !== undefined && (typeof location.elevation !== 'number' || location.elevation < -500)) {
    throw new PanchangError(
      `Elevation must be >= -500 meters, got ${location.elevation}`,
      'INVALID_ELEVATION'
    );
  }
}

export function validateDate(date: Date): void {
  if (!(date instanceof Date) || isNaN(date.getTime())) {
    throw new PanchangError(`Invalid Date: ${String(date)}`, 'INVALID_DATE');
  }
  const year = date.getFullYear();
  if (year < 1900 || year > 2100) {
    throw new PanchangError(
      `Date must be between 1900 and 2100 for astronomical accuracy, got year ${year}`,
      'INVALID_DATE'
    );
  }
}

// ── Programmer-error guards for internal cyclic indices ──────────────────
// These throw `RangeError` (programmer errors at module boundaries), distinct
// from the user-facing `PanchangError` thrown by validateDate / validateLocation.

function assertCyclicIndex(value: number, modulus: number, name: string): void {
  if (!Number.isInteger(value) || value < 0 || value >= modulus) {
    throw new RangeError(`${name} must be integer in [0, ${modulus - 1}], got ${value}`);
  }
}

export function assertNakshatraIndex(value: number, name = 'nakshatra index'): void {
  assertCyclicIndex(value, TOTAL_NAKSHATRAS, name);
}

export function assertVaraIndex(value: number, name = 'vara index'): void {
  assertCyclicIndex(value, 7, name);
}

export function assertTithiIndex(value: number, name = 'tithi index'): void {
  assertCyclicIndex(value, TOTAL_TITHIS, name);
}

File: source/go/internal/utils/validation.go
package utils

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/types"
)

func ValidateLocation(location types.GeoLocation) error {
	if math.IsNaN(location.Latitude) || math.IsInf(location.Latitude, 0) ||
		location.Latitude < -90 || location.Latitude > 90 {
		return types.Codef(types.ErrInvalidLatitude,
			"Latitude must be a finite number between -90 and 90, got %s",
			jsString(location.Latitude))
	}
	if math.IsNaN(location.Longitude) || math.IsInf(location.Longitude, 0) ||
		location.Longitude < -180 || location.Longitude > 180 {
		return types.Codef(types.ErrInvalidLongitude,
			"Longitude must be a finite number between -180 and 180, got %s",
			jsString(location.Longitude))
	}
	if math.IsNaN(location.Elevation) || math.IsInf(location.Elevation, 0) ||
		location.Elevation < -500 {
		return types.Codef(types.ErrInvalidElevation,
			"Elevation must be >= -500 meters, got %s", jsString(location.Elevation))
	}
	return nil
}

func ValidateDate(ms int64) error {
	year := utcYear(ms)
	if year < 1900 || year > 2100 {
		return types.Codef(types.ErrInvalidDate,
			"Date must be between 1900 and 2100 for astronomical accuracy, got year %d", year)
	}
	return nil
}

func assertCyclicIndex(value, modulus int, name string) error {
	if value < 0 || value >= modulus {
		return types.Codef(types.ErrInvalidInput,
			"%s must be integer in [0, %d], got %d", name, modulus-1, value)
	}
	return nil
}

func AssertNakshatraIndex(value int, name string) error {
	return assertCyclicIndex(value, TotalNakshatras, defaultName(name, "nakshatra index"))
}

func AssertVaraIndex(value int, name string) error {
	return assertCyclicIndex(value, 7, defaultName(name, "vara index"))
}

func AssertTithiIndex(value int, name string) error {
	return assertCyclicIndex(value, TotalTithis, defaultName(name, "tithi index"))
}

func defaultName(given, fallback string) string {
	if given == "" {
		return fallback
	}
	return given
}
